don't use canceled context to send KILL signal to healthcheck process

ndeloof opened the pull request:
**- What I did**
send KILL signal to healthcheck process when the context is canceled (timeout)
- closes #43737

**- How I did it**
Use a background context to send KILL signal

**- How to verify it**
see reproduction scenario on #43737

**- Description for the changelog**
fix healthcheck process termination on timeout

**- A picture of a cute animal (not mandatory but encouraged)**
![image](https://user-images.githubusercontent.com/132757/175269858-984e8149-cf45-4d12-9380-831b261cfda3.png)

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
do we want a test for this? Or too complicated to set one up?

thaJeztah commented:
I don't think the failure was related, so I kicked CI again, but posting here in case there's a flaky test at hand that we're not tracking yet;

```
=== RUN   TestDockerSwarmSuite/TestAPISwarmServicesMultipleAgents
    docker_api_swarm_service_test.go:100: [d7f46e9e54ee1] joining swarm manager [d451e1c460ace]@0.0.0.0:2477, swarm listen addr 0.0.0.0:2478
    docker_api_swarm_service_test.go:101: [d285f6cd0aebd] joining swarm manager [d451e1c460ace]@0.0.0.0:2477, swarm listen addr 0.0.0.0:2479
    docker_api_swarm_service_test.go:122: timeout hit after 30s: output: "4adc199e5240\n13ecfed699f5\n", output: "5547cb8d033b\n97e5710958e7\nb9aa8d4c5a86\n3440eeecbb60\n8fabde17fa91\n"
    poll.go:125: assertion failed: error is not nil: exit status 1
    check_test.go:611: [d7f46e9e54ee1] daemon is not started
    --- FAIL: TestDockerSwarmSuite/TestAPISwarmServicesMultipleAgents (60.05s)
```

ndeloof commented on integration/container/health_test.go:
```diff
@@ -93,6 +93,32 @@ while true; do sleep 1; done
+
+	cID := container.Run(ctx, t, client, func(c *container.TestContainerConfig) {
+		c.Config.Healthcheck = &containertypes.HealthConfig{
+			Test:     []string{"CMD-SHELL", "sleep 30"},
```
unfortunately, healthcheck output when interrupted is not captured to be reported by inspect, and replaced by a gerenic timeout message:
```json
            "Health": {
                "Status": "unhealthy",
                "Log": [
                    {
                        "Start": "2022-06-24T08:59:46.200205085Z",
                        "End": "2022-06-24T08:59:47.203354127Z",
                        "ExitCode": -1,
                        "Output": "Health check exceeded timeout (1s)"
                    },
```
This would require the `exec` logic to change, so that it captures the killed process output even in such circumstances. Would this be fine?

thaJeztah replied:
If the question is; is it ok to add addition text to `Output`,  good question; I'm not aware of any code parsing the output; AFAIK, it's intended purely as "additional info", but the description is a bit ambiguous; https://github.com/moby/moby/blob/c0069b83bce03d1dd1ac2c04bd309f0641bc8114/api/types/types.go#L400


Do we capture the output if the HealthCheck _is_ successful? (e.g. if a successful health check would print `OK`, is that output returned?)

So we could either;

- include both: `Health check exceeded timeout (1s): <output>`
- Or _only_ include `<output>` (which would be a closer match to the description, but I guess the timeout information is still useful

Perhaps we need to update the description of the field to better clarify what the expectation is;

- for success
- on error (is it the raw error output? other output?)

cpuguy83 replied:
Oh right this could be more difficult than just capturing the output because we won't know when the command is completed (or if it ever will).

thaJeztah commented on integration/container/testdata/health/go.mod:
```diff
@@ -0,0 +1,3 @@
+module github.com/docker/docker/integration/container/testdata/health
+
+go 1.17
```
This one is no longer needed, correct?

thaJeztah commented:
Something failing;

```
=== RUN   TestHealthCheckProcessKilled
    health_test.go:111: timeout hit after 10s: waiting for a new container healthcheck
--- FAIL: TestHealthCheckProcessKilled (11.40s)
```

thaJeztah commented on daemon/health.go:
```diff
@@ -136,9 +136,16 @@ func (p *cmdProbe) run(ctx context.Context, d *Daemon, cntr *container.Container
+		out := output.String()
+		var msg string
+		if len(out) > 0 {
+			msg = fmt.Sprintf("Health check exceeded timeout (%v):\n%s", probeTimeout, out)
```
Before I forget to ask; I see this adds a newline before the output; was there a specific reason for that? Or was it more readable if the output contained multiple lines (e.g.)?

Was wondering if it'd be more readable (assuming single line output) to have it on the same line (`..exceeded timeout(): <output>`) to look more like a regular error message, but perhaps that didn't work well, so thought I'd ask

ndeloof replied:
no specific reason, feels more natural for me to add timeout message as a separate line, then command output

corhere commented on daemon/exec.go:
```diff
@@ -292,16 +293,15 @@ func (daemon *Daemon) ContainerExecStart(ctx context.Context, name string, optio
+		case <-sigCtx.Done():
 			logrus.Infof("Container %v, process %v failed to exit within %v of signal TERM - using the force", c.ID, name, termProcessTimeout)
-			daemon.containerd.SignalProcess(ctx, c.ID, name, signal.SignalMap["KILL"])
+			daemon.containerd.SignalProcess(context.Background(), c.ID, name, signal.SignalMap["KILL"])
```
`SignalProcess` sends an RPC, therefore could potentially hang for a long time if things go wrong. Some timeout might be appropriate?

I wish execs were terminated by calling a method rather than by canceling a context. A caller would then be able to pass a dedicated context just for the termination operation instead of having to hardcode some policy into the `ContainerExecStart` implementation.

ndeloof replied:
Makes sense to me to use a timeout here, but which value to use? `termProcessTimeout` applies to target process to terminate, not internal communication delays.

About global design, I agree use of a cancelable context here to stop exec is not very pleasant, considering the significant amount of additional API calls involved.

corhere replied:
I'd pick a timeout value which is large enough to never be hit when containerd is responsive, accounting for potentially massive network latencies, but which is not actually forever so that requests to an unresponsive containerd (or a [slowloris](https://en.wikipedia.org/wiki/Slowloris_(computer_security)) honeypot masquerading as a containerd) don't leak resources indefinitely. How does 30 seconds each sound?

ndeloof commented:
I'm stuck here with test failing on CI
`health_test.go:112: timeout hit after 10s: waiting for container healthcheck logs`

vvoland commented on daemon/exec.go:
```diff
@@ -293,15 +296,14 @@ func (daemon *Daemon) ContainerExecStart(ctx context.Context, name string, optio
-		daemon.containerd.SignalProcess(ctx, c.ID, name, signal.SignalMap["TERM"])
+		daemon.containerd.SignalProcess(context.Background(), c.ID, name, signal.SignalMap["TERM"])
 
 		timeout := time.NewTimer(termProcessTimeout)
```
Process has 10s to process the SIGTERM - in the test case SIGTERM is just ignored, so this will have to wait 10 seconds.

ndeloof replied:
Oh, so obvious I didn't saw it... thanks

vvoland commented on integration/container/health_test.go:
```diff
@@ -93,6 +94,42 @@ while true; do sleep 1; done
+			Retries:  1,
+		}
+	})
+	poll.WaitOn(t, pollForHealthCheckLog(ctx, client, cID, "Health check exceeded timeout (50ms):\nSIGTERM\n"))
```
And the default timeout for poll.WaitOn is 10s :) So the SIGKILL isn't sent, because the test ends first.
This could be fixed by increasing the timeout here, but that means that the test will take 10s to finish... 🤔

corhere commented on integration/container/health_test.go:
```diff
@@ -93,6 +94,41 @@ while true; do sleep 1; done
+
+	cID := container.Run(ctx, t, client, func(c *container.TestContainerConfig) {
+		c.Config.Healthcheck = &containertypes.HealthConfig{
+			Test:     []string{"CMD-SHELL", "sh -c 'trap \"echo SIGTERM\" 15; while true; do true; done;'"},
```
I'm not so sure it's that `sleep` is blocking signals. I suspect that what's actually happening is busybox `ash` waits for the `sleep` subprocess to exit before executing the trap, [matching the behaviour of `bash`](https://www.gnu.org/software/bash/manual/html_node/Signals.html#:~:text=If%20Bash%20is,trap%20is%20executed.), and the daemon is only sending the signals to the topmost shell process so the `sleep` subprocess is never sent a signal.

Maybe something like this would work?
```sh
trap "echo SIGTERM" 15
sleep 30 &
pid=$!
while jobs %% >/dev/null; do wait $pid; done
```
I tested this interactively in `library/busybox:latest` and it seems to work, though the background job leaks.

Although, if it really is the case that the trap is not executing because the shell is waiting for the `sleep` command to complete, running `usleep` in a loop would effectively "poll" for the trapped signals with an upper-bound latency of the duration of the `usleep` command. No fancy stuff with background jobs required!
```suggestion
			Test:     []string{"CMD-SHELL", "sh -c 'trap \"echo SIGTERM\" 15; while true; do usleep 50000; done;'"},
```

thaJeztah commented:
@ndeloof @corhere while reviewing, I was playing a bit with some options; I opened them as https://github.com/ndeloof/docker/pull/2

First commit of that branch moves the timeout to ExecConfig
Second commit changes the timeout for health-checks to a fixed (but short) value; I picked 0.5 seconds, but open to discussion for that.

Feedback / thoughts welcome 😅

corhere commented on daemon/exec.go:
```diff
@@ -293,15 +296,22 @@ func (daemon *Daemon) ContainerExecStart(ctx context.Context, name string, optio
+		daemon.containerd.SignalProcess(sigCtx, c.ID, name, signal.SignalMap["TERM"])
 
-		timeout := time.NewTimer(termProcessTimeout)
+		wait := termProcessTimeout
+		if ec.StopTimeout != nil {
+			wait = *ec.StopTimeout
+		}
```
If we're making the timeout configurable anyway, what do you think of dropping the arbitrary default value and leave it entirely up to the caller to dictate the policy? I'm thinking along the lines of [mirroring this proposal for `os/exec`](https://github.com/golang/go/issues/50436) which defaults to insta-kill on cancel while allowing for both the graceful-terminate signal and timeout to be configured.

corhere replied:
To clarify, I am not suggesting we expose any features to users. My line of thinking is that the API handler `func (*containerRouter) postContainerExecCreate` would be a more appropriate place to dictate the stop-timeout policy for execs created via the API, even if it is the existing hardcoded `termProcessTimeout` policy, just like how `cmdProbe` now dictates the policy for health-check execs. The obvious implementation would require adding a `StopTimeout` field to `types.ExecConfig`, though tagging it <tt>\`json:"-"\`</tt> would keep it from being exposed in the Engine API, just like its analogue in `container.HealthConfig`.

thaJeztah replied:
Not 100% sure if those defaults should make its way into the API endpoint, but perhaps in `Daemon.ContainerExecCreate()`, as that's where the exec config is created (and to match what I did for the health check) https://github.com/moby/moby/blob/4eb1c5bd52afa08595d9a082a9ba05c81f79be07/daemon/exec.go#L98

That way the code here (`ContainerExecStart`) wouldn't have to bother with checking if all options are there, and just consume the exec-config as stored. Perhaps make `exec.NewConfig()` accept options so that we don't have the awkward code in both locations to first do a `exec.NewConfig()`, and then having to manually manipulate it to set all the other options.

Given that the health check code doesn't use the endpoint, nor `Daemon.ContainerExecCreate()`, I think that would also be fine to do separately though 🤔 WDYT?

ndeloof replied:
I totally agree default values should be handled while initializing struct, vs checking for nil to apply defaults.
Constructors is one of the language feature I miss the most writing Go. I'd be happy we rely on `NewXX` initializer pattern everywhere possible

thaJeztah replied:
> I totally agree default values should be handled while initializing struct, vs checking for nil to apply defaults.

I should note that it's not always possible to do so when constructing things; some default should be "baked" into a struct, whereas some other defaults may have to be evaluated at runtime (e.g. if option X isn't set, take the current default as configured in the environment and/or daemon configuration).

But I agree that having a canonical location to where defaults are set (_where possible_) is good. Currently there's too many places where defaults are set, which means you often have to drill through a lot of code to figure out where they're set.

thaJeztah replied:
^^ we should not increase the scope of this PR too much though; I haven't checked yet what would be needed, but I'd still be fine to tackle that in a follow-up (more "improve exec in general, not just for health-check").

Perhaps just moving this to `ContainerExecCreate()` would work though;

```go
if ec.StopTimeout != nil {
	wait = *ec.StopTimeout
}
```

Which could be just a line added here; https://github.com/moby/moby/blob/4eb1c5bd52afa08595d9a082a9ba05c81f79be07/daemon/exec.go#L128


```go
execConfig.StopTimeout = termProcessTimeout
```

thaJeztah commented on daemon/health.go:
```diff
@@ -81,6 +81,8 @@ func (p *cmdProbe) run(ctx context.Context, d *Daemon, cntr *container.Container
 	execConfig.Privileged = false
 	execConfig.User = cntr.Config.User
 	execConfig.WorkingDir = cntr.Config.WorkingDir
+	// health checks don't get the full 10 seconds timeout for a .
```
whoops; looks like I was still writing here but didn't finish that before opening https://github.com/ndeloof/docker/pull/2

Trying to recall what I wanted to write here; I guess something about stopping gracefully.

```go
// health checks don't get the full 10 seconds timeout to stop gracefully.
```

thaJeztah commented on daemon/health.go:
```diff
@@ -81,6 +81,11 @@ func (p *cmdProbe) run(ctx context.Context, d *Daemon, cntr *container.Container
 	execConfig.WorkingDir = cntr.Config.WorkingDir
+	// health checks don't get the full 10 seconds timeout to stop gracefully.
+	execConfig.StopTimeout = 500 * time.Millisecond
+	if cntr.Config.Healthcheck.StopTimeout != 0 {
+		execConfig.StopTimeout = cntr.Config.Healthcheck.StopTimeout
+	}
```
Wondering why you added this back in your last push; is this only to make the test run faster (50 ms instead of 500)? Otherwise we can omit the changes to the `Healthcheck` type

thaJeztah replied:
Gave it a quick try, and it seems to work, but let me know if you had a specific reason for this 😅 https://github.com/ndeloof/docker/pull/3

ndeloof replied:
yes, goal is to keep test fast. Not being able to control timing demonstrate lack of design for testability imho

thaJeztah replied:
Not everything has to be configurable; we could make the fixed value shorter of course; not sure if saving half a second for a test warrants a change to an exported struct though (if it's only used in a test)

corhere commented:
The code to terminate health check execs gracefully has been broken [since v17.11](https://github.com/moby/moby/commit/ddae20c032058a0fd42c34c2e9750ee8f6296ac8#diff-73eee5ff6da257cc3f90a10ff07f4d47e02e7999e317f4ce0bb9eab7f662193fR263), nearly five years ago! The "statute of limitations" has arguably expired on the feature and so we have an opportunity to change the behaviour. The graceful termination code in `ContainerExecStart` is only used by health checks (the exec API handler calls `ContainerExecStart` with the background context) so we could replace it with whatever other behaviour we wanted. Kubernetes exec probes with cri-containerd send `SIGKILL` immediately upon timeout; we could do the same. That would both simplify the implementation and speed up test runs by half a second. @thaJeztah mentioned during today's maintainer meeting that he might take that on in an alternative PR so you may want to with him to avoid duplicating work.

olljanat reviewed (approved):
I can confirm that issue #43737 does not exist on Windows with HCS v1 but will appear there when containerd mode is enabled and this fix looks working fine on there too.

LGTM

corhere commented:
@thaJeztah @cpuguy83 @tianon PTAL

thaJeztah commented:
CRIU repo down again? Let me give it another kick

```
failed to load cache key: Get "https://download.opensuse.org/repositories/devel:/tools:/criu/Debian_11/Release.key": dial tcp 195.135.221.134:443: i/o timeout
script returned exit code 1
```

thaJeztah commented on daemon/health.go:
```diff
@@ -133,12 +133,19 @@ func (p *cmdProbe) run(ctx context.Context, d *Daemon, cntr *container.Container
+		// Wait for probe to exit (it might take some time to call containerd to kill
+		// the process and we don't want dying probes to pile up).
 		<-execErr
+		out := output.String()
+		var msg string
+		if len(out) > 0 {
```
Oh, since you need to update for the other bit; perhaps also make the `out` variable local here (slightly clearer to indicate it's not used elsewhere);
```suggestion

		var msg string
		if out := output.String(); len(out) > 0 {
```

thaJeztah replied:
👇 oh, and perhaps you have thoughts on the newline as well; I'm not _really_ tried to it, mostly wondering if using a space instead of a newline would be more "standard" (and possibly ease grep-ing for failed checks combined with output - not sure)

thaJeztah reviewed (changes requested):
@corhere looks like there's a small issue (pointed out); also left two suggestions for consideration.

Do we want to keep the commits separate? (mostly looking at the test, which is only added in the second commit 😅

thaJeztah commented:
> mostly looking at the test, which is only added in the second commit

I think the `output` was added for the original change where we needed the output to see if the trap was called; from that perspective, it's not needed anymore to test the fix, so I _think_ we could have a basic test in the first commit (if we want to backport to 20.10), and make the second commit an "enhancement" commit (to include the output), which could be in 22.06; WDYT?

thaJeztah commented:
@corhere I pushed a branch making that split; let me know if that looks good to you (if so, I could push to this PR); https://github.com/moby/moby/compare/master...thaJeztah:healthcheck_timeout_split?expand=1

corhere commented:
@thaJeztah your branch LGTM; push away! I have no strong feelings either way regarding the newline in the timeout msg.

thaJeztah commented:
pushed 👍

thaJeztah commented:
hm... did I miss a change on Windows? It's passing on Linux, but fails on Windows;

```
=== Failed
=== FAIL: github.com/docker/docker/integration/container TestHealthCheckProcessKilled (3.23s)
    health_test.go:111: polling check failed: expected "Health check exceeded timeout (50ms): logs logs logs\n", got "Health check exceeded timeout (50ms)"
```

Let me check 🤔

thaJeztah commented:
I guess `output.String()` is empty on Windows for some reason.

Given that (as mentioned above) capturing the logs is not essential for the fix, perhaps I should just move it to a separate follow-up PR; let me do so

thaJeztah commented:
This is weird; it's as if Windows is running an older commit before the quote was fixed;

```
=== FAIL: github.com/docker/docker/integration/container TestHealthCheckProcessKilled (unknown)
    health_test.go:111: polling check failed: expected "Health check exceeded timeout (50ms)", got "60': line 1: syntax error: unterminated quoted string\n"
coverage: [no statements]
panic: test timed out after 1h0m0s
```

At least, I don't see anything wrong looking at this 🤔;

```go
Test:     []string{"CMD-SHELL", `sh -c 'sleep 60'`},
```

thaJeztah commented:
I guess `CMD-SHELL` already runs with a shell, so we can probably omit the `sh -c` 🤔

thaJeztah commented:
```
github.com/docker/docker/integration/container TestHealthCheckProcessKilled (16.06s)
    main_test.go:32: assertion failed: error is not nil: Error response from daemon: Could not kill running container 8147d85d217100444d4d50c351608743d6448fd8f9b200cc1a9491a4d83a1b32, cannot remove - tried to kill container, but did not receive an exit event: failed to remove 8147d85d217100444d4d50c351608743d6448fd8f9b200cc1a9491a4d83a1b32

thaJeztah commented on integration/container/health_test.go:
```diff
@@ -93,6 +94,44 @@ while true; do sleep 1; done
+	})
+	poll.WaitOn(t, pollForHealthCheckLog(ctx, apiClient, cID, "Health check exceeded timeout (50ms)"))
+
+	err := apiClient.ContainerKill(ctx, cID, "SIGKILL")
+	assert.NilError(t, err)
```
Let's try if this helps with the cleanup step on Windows 🤞

thaJeztah commented:
https://github.com/moby/moby/pull/43739#discussion_r950687752

> Let's try if this helps with the cleanup step on Windows 🤞

Unfortunately; that didn't help, only shifting / repeating the problem;

```
=== FAIL: github.com/docker/docker/integration/container TestHealthCheckProcessKilled (27.95s)
    health_test.go:114: assertion failed: error is not nil: Error response from daemon: Cannot kill container: d5e1f84f4dbb83aac2623d972592089575ed8a3917e9d8ddaa487b423938c461: tried to kill container, but did not receive an exit event
    main_test.go:32: assertion failed: error is not nil: Error response from daemon: Could not kill running container d5e1f84f4dbb83aac2623d972592089575ed8a3917e9d8ddaa487b423938c461, cannot remove - tried to kill container, but did not receive an exit event: failed to remove d5e1f84f4dbb83aac2623d972592089575ed8a3917e9d8ddaa487b423938c461
```

So it looks like on Windows, the health check causes an issue with stopping the container (?). Of course the question is; was that already the case (but not tested before?), or is there an issue in the implementation here? (Perhaps an issue with signal handling / incorrect signals for Windows, idk)

thaJeztah commented on integration/container/health_test.go:
```diff
@@ -93,6 +94,44 @@ while true; do sleep 1; done
+
+	cID := container.Run(ctx, t, apiClient, func(c *container.TestContainerConfig) {
+		c.Config.Healthcheck = &containertypes.HealthConfig{
+			Test:     []string{"CMD", "sleep", "60"},
```
Other attempt; trying `CMD ["sleep", "60"]` instead of starting `sleep` in a shell.

thaJeztah commented:
However, on Windows 2019 and Windows 2022 running with containerd, `TestHealthCheckProcessKilled` times out; wondering if this could be related to something like https://github.com/containerd/containerd/issues/5067

```
=== Failed
=== FAIL: github.com/docker/docker/integration/container TestHealthCheckProcessKilled (unknown)
coverage: [no statements]
panic: test timed out after 1h0m0s

goroutine 199 [running]:
testing.(*M).startAlarm.func1()
	C:/hostedtoolcache/windows/go/1.18.5/x64/src/testing/testing.go:2029 +0x8e
created by time.goFunc
	C:/hostedtoolcache/windows/go/1.18.5/x64/src/time/sleep.go:176 +0x32
	...
```

corhere commented on integration/container/health_test.go:
```diff
@@ -93,6 +94,44 @@ while true; do sleep 1; done
+
+	cID := container.Run(ctx, t, apiClient, func(c *container.TestContainerConfig) {
+		c.Config.Healthcheck = &containertypes.HealthConfig{
+			Test:     []string{"CMD-SHELL", `sh -c "sleep 60"`},
```
Relevant CI logs: [builtin](https://github.com/moby/moby/runs/7949749461?check_suite_focus=true#step:21:2952), [`containerd-shim-runhcs-v1`](https://github.com/moby/moby/runs/7949748914?check_suite_focus=true#step:19:2368)

In both cases, the `CommandLine` string ultimately passed into the Win32API `CreateProcess` is
```
cmd /S /C "sh -c \"sleep 60\""
```

After lots of research and experimentation on a Windows box... my head hurts. Here's what I know so far:
- A Windows process' command and arguments are passed in as a single string, which is the string specified in the `CommandLine` passed into `CreateProcess`.
- Each process parses its own command line. Not all processes use the same algorithm.
  - As a corollary,`cmd.exe` does not tokenize its subprocess' command lines for them.
- `cmd.exe /S /C` follows some arcane rules, but the gist is that everything following the `/C` is taken literally as a command line, with one exception: if the first non-whitespace character following the `/C` is a double-quote character, the command line is modified by dropping that character and the last double-quote character in the remaining command-line.

Putting that all together:
`cmd /S /C "sh -c \"sleep 60\""` is equivalent to interactively running
```
C:\> sh -c \"sleep 60\"
```
Executing that command results in `CreateProcess()` being called with `CommandLine` set to `sh -c \"sleep 60\"`.

The Windows C Runtime parses the `CommandLine` `sh -c \"sleep 60\"` into the argv
```
argv[0] = sh
argv[1] = -c
argv[2] = \"sleep
argv[3] = 60\"
```
and busybox ash gets very confused.

The ultimate issue is that `cmdProbe` is incorrectly escaping the shell command line in a manner which the Windows C runtime would parse as a single argument, expecting `cmd.exe /S /C` to act like `sh -c` and consume the next argv as its argument. But `cmd.exe` does not unescape the argument.

https://github.com/containerd/containerd/issues/5067 is very related to the above, but I don't think it has to do with the timeouts you saw. (My guess is that the timeouts have something to do with `sleep.exe` not existing.)

thaJeztah replied:
I recall I tested various combinations; any of them would fail on at least _one_ combination 😞 

I tried running `sleep` as main process itself;

```
[]string{"CMD", "sleep",  "60"},
```

Which would be the equivalent of the default command we set for "sleep" containers; https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/integration-cli/test_vars_test.go#L6-L9

Tried running it as child process of `sh`;

```
[]string{"CMD", "sh", "-c", "sleep 60"},
```

Tried passing it as single string, but I think Linux wasn't a fan of that

```
[]string{"CMD-SHELL", `sh -c "sleep 60`},
```


And the original one (which (like above), and if I'm not mistaken, would run `sh` as child process of `cmd`, so that didn't feel right);

```
[]string{"CMD-SHELL", "sh", "-c", "sleep 60"},
```

thaJeztah replied:
One thing I wonder is if the Busybox image should also be updated to set the `SHELL` to `/bin/sh`; currently it doesn't, which means that (in Dockerfiles), the default shell would still be `cmd` or `powershell` (depending on the base image)

thaJeztah replied:
Also wondering if `exec.Config` needs to have a `CommandLine` property added; it just occurred to me that we have that for the container's main process, but don't have it for `exec` 🤔 https://github.com/moby/moby/blob/2646beac875eaf69930387d04fceca89e93d1da1/vendor/github.com/opencontainers/runtime-spec/specs-go/config.go#L41-L43

corhere replied:
> One thing I wonder is if the Busybox image should also be updated to set the `SHELL` to `/bin/sh`

> Also wondering if `exec.Config` needs to have a `CommandLine` property added

Both excellent ideas I wholeheartedly support!

corhere committed:
daemon: kill exec process on ctx cancel

Terminating the exec process when the context is canceled has been
broken since Docker v17.11 so nobody has been able to depend upon that
behaviour in five years of releases. We are thus free from backwards-
compatibility constraints.

Co-authored-by: Nicolas De Loof <nicolas.deloof@gmail.com>
Co-authored-by: Sebastiaan van Stijn <github@gone.nl>
Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>
Signed-off-by: Cory Snider <csnider@mirantis.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
windows 2022 / containerd;

```
=== FAIL: github.com/docker/docker/integration/container TestHealthCheckProcessKilled (15.43s)
    main_test.go:32: assertion failed: error is not nil: Error response from daemon: Could not kill running container d4a736c61017160c76188113089560575c359d42db1ff5b6e2e18838002a985d, cannot remove - tried to kill container, but did not receive an exit event: failed to remove d4a736c61017160c76188113089560575c359d42db1ff5b6e2e18838002a985d

thaJeztah commented on integration/container/health_test.go:
```diff
@@ -93,6 +94,42 @@ while true; do sleep 1; done
 
+// TestHealthCheckProcessKilled verifies that health-checks exec get killed on time-out.
+func TestHealthCheckProcessKilled(t *testing.T) {
+	skip.If(t, testEnv.RuntimeIsWindowsContainerd(), "FIXME: Broken on Windows + containerd combination")
```
I added a `skip` for now.

thaJeztah commented:
Some unrelated failures; not sure I've seen this one before, but doesn't look related at least (kicked CI once more)

```
=== FAIL: github.com/docker/docker/integration-cli TestDockerCLIPsSuite/TestPsListContainersFilterID (19.48s)
    check_test.go:402: assertion failed: error is not nil: Error response from daemon: Could not kill running container a617f7b8cf70de3b92c7e740bcd129fba04c00d86457b93fb434379dc9b7cae5, cannot remove - tried to kill container, but did not receive an exit event: failed to remove a617f7b8cf70de3b92c7e740bcd129fba04c00d86457b93fb434379dc9b7cae5
    --- FAIL: TestDockerCLIPsSuite/TestPsListContainersFilterID (19.48s)
```

thaJeztah reviewed (approved):
It's greeeen!

LGTM

thaJeztah commented:
Alright, let's get this one in
